presto-main/src/test/java/io/prestosql/tests/LogTestDurationListener.java

```diff
@@ -13,30 +13,121 @@
  */
 package io.prestosql.tests;
 
+import com.google.common.collect.ImmutableMap;
 import io.airlift.log.Logger;
 import io.airlift.units.Duration;
 import org.testng.IClassListener;
+import org.testng.IExecutionListener;
 import org.testng.IInvokedMethod;
 import org.testng.IInvokedMethodListener;
 import org.testng.ITestClass;
 import org.testng.ITestResult;
 
+import javax.annotation.concurrent.GuardedBy;
+
+import java.lang.management.ThreadInfo;
+import java.util.Arrays;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
 
 import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.concurrent.Threads.daemonThreadsNamed;
 import static io.airlift.units.Duration.nanosSince;
 import static java.lang.String.format;
+import static java.lang.management.ManagementFactory.getThreadMXBean;
+import static java.util.stream.Collectors.joining;
 
 public class LogTestDurationListener
-        implements IClassListener, IInvokedMethodListener
+        implements IExecutionListener, IClassListener, IInvokedMethodListener
 {
     private static final Logger LOG = Logger.get(LogTestDurationListener.class);
 
     private static final Duration SINGLE_TEST_LOGGING_THRESHOLD = Duration.valueOf("30s");
     private static final Duration CLASS_LOGGING_THRESHOLD = Duration.valueOf("1m");
+    private static final Duration GLOBAL_IDLE_LOGGING_THRESHOLD = Duration.valueOf("5m");
+
+    private final ScheduledExecutorService scheduledExecutorService;
 
     private final Map<String, Long> started = new ConcurrentHashMap<>();
+    private final AtomicLong lastChange = new AtomicLong(System.nanoTime());
+    private final AtomicBoolean hangLogged = new AtomicBoolean();
+    private final AtomicBoolean finished = new AtomicBoolean();
+    @GuardedBy("this")
+    private ScheduledFuture<?> monitorHangTask;
+
+    public LogTestDurationListener()
+    {
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(daemonThreadsNamed("TestHangMonitor"));
+    }
+
+    @Override
+    public synchronized void onExecutionStart()
+    {
+        resetHangMonitor();
+        finished.set(false);
```

Comment:
is necessary?

BTW this class may behave incorrectly when using XML testng file with multiple suite/test sections.
I think this is not something we need to fix today, just noting.

Reply:
This let's us print a message saying that everything finished, but the JVM didn't shutdown, that way when there is a problem we know we just need to look for the non-daemon threads.

Reply:
The listener interface is documented as triggering before and after all TestNG stuff